src/gadgets/bits2num.rs

```diff
@@ -0,0 +1,270 @@
+    pub(crate) selector: Selector,
+    pub(crate) input: Column<Advice>,
+    pub(crate) output: Column<Advice>,
+}
+
+pub(crate) struct Bits2NumChip<F: Field> {
+    config: Bits2NumChipConfig,
+    _marker: PhantomData<F>,
+}
+
+impl<F: PrimeField> Bits2NumChip<F> {
+    pub(crate) fn construct(config: Bits2NumChipConfig) -> Self {
+        Bits2NumChip {
+            config,
+            _marker: PhantomData::default(),
+        }
+    }
+
+    pub(crate) fn configure(
+        meta: &mut ConstraintSystem<F>,
+        input: Column<Advice>,
+        output: Column<Advice>,
+    ) -> Bits2NumChipConfig {
+        let selector = meta.selector();
+
+        meta.create_gate("next_num_constraint", |cs| {
+            let bit_val = cs.query_advice(input, Rotation::cur());
+            let prev_num_val = cs.query_advice(output, Rotation::cur());
+            let num_val = cs.query_advice(output, Rotation::next());
+
+            let selector = cs.query_selector(selector);
+
+            Constraints::with_selector(
+                selector,
+                vec![num_val - (prev_num_val * F::from(2) + bit_val)],
+            )
+        });
+
+        Bits2NumChipConfig {
+            selector,
+            input,
+            output,
+        }
+    }
+}
+
+impl<F: PrimeField> Bits2NumInstruction<F> for Bits2NumChip<F> {
+    fn convert_be(
+        &self,
+        layouter: &mut impl Layouter<F>,
+        bits: Vec<AssignedCell<F, F>>,
+    ) -> Result<AssignedCell<F, F>, Error> {
+        let res = layouter.assign_region(
+            || "bits2num",
+            |mut region| {
+
+                let num_bit_size = bits.len();
+                assert!(
+                    num_bit_size as u32 <= F::CAPACITY,
+                    "Number of bits is too large for field size!"
+                );
+
+                let mut num_val = Value::known(F::from(0));
+
+                let mut num_val_cell = region.assign_advice_from_constant(
+                    || format!("prev_num_val {}", 0),
+                    self.config.output,
+                    0,
+                    F::ZERO,
+                )?;
+
+                for i in 0..num_bit_size {
+                    self.config.selector.enable(&mut region, i).unwrap();
+
+                    num_val = num_val * Value::known(F::from(2)) + bits[i].value();
+
+                    num_val_cell = region.assign_advice(
+                        || format!("num_val {}", i + 1),
+                        self.config.output,
+                        i + 1,
+                        || num_val,
+                    )?;
+
+                    bits[i].copy_advice(
+                        || format!("input bit {}", i),
+                        &mut region,
+                        self.config.input,
+                        i,
+                    )?;
+                }
+
+                Ok(num_val_cell)
+            },
+        );
+
+        res
+    }
+
+    fn convert_le(
```

Comment:
I like this pattern, very clean! That way, there is no confusion about how the gadget works.